File: scripts/edmontonMatch.R
# edmontonMatch.R
# Match building permits to assessment
# Tom Davidoff
# 07/16/26

library(data.table)
library(fixest)
library(sf)
library(duckdb)
library(ggplot2)

assessorHistoricalPath <- "~/DropboxExternal/dataRaw/edmonton/Property_Assessment_Data_(Historical)_20260611.csv"
assessorCurrentPath <- "~/DropboxExternal/dataRaw/edmonton/Property_Information_(Current_Calendar_Year)_20260612.csv"

dtAH <- fread(assessorHistoricalPath,select=c("Lot Size","Account Number","Zoning","Legal Description","Latitude","Longitude","Assessment Year","Assessed Value"))
dtAC <- fread(assessorCurrentPath,select = c("lot_size","zoning","legal_description","Latitude","Longitude","Account Number","year_built","Total Gross Area"))
dtAC <- dtAC[zoning=="RS"]
setkey(dtAH, "Account Number")
setkey(dtAC, "Account Number")
dtAM <- dtAH[dtAC]

# permit match
permitPath <- "~/DropboxExternal/dataRaw/edmonton/edmontonBuildingPermits.csv"
con <- dbConnect(duckdb())
q <- sprintf(R"(
  SELECT
    "PERMIT_DATE", "PERMIT_NUMBER", "YEAR", "BUILDING_TYPE", "WORK_TYPE",
    "CONSTRUCTION_VALUE", "FLOOR_AREA", "UNITS_ADDED", "ADDRESS",
    "LEGAL_DESCRIPTION", "ZONING", "NEIGHBOURHOOD", "LATITUDE", "LONGITUDE"
  FROM read_csv_auto('%s')
  WHERE ZONING = 'RS'
)", permitPath)
dfP <- dbGetQuery(con,q)
dbDisconnect(con, shutdown = TRUE)
dtP <- as.data.table(dfP)
dtP[,isRowHouse := grepl("Row House",BUILDING_TYPE)]
dtP[,isDetached := grepl("Single Detached",BUILDING_TYPE)]

# only new construction or construction value above critical percentile of new building types
QCRIT <- .25
qNew <- dtP[isRowHouse | isDetached,quantile(CONSTRUCTION_VALUE/FLOOR_AREA,probs=c(QCRIT),na.rm=TRUE)]
dtP <- dtP[isRowHouse | isDetached | (CONSTRUCTION_VALUE/FLOOR_AREA)>=qNew]

dtAM[,LEGAL_DESCRIPTION := gsub("\\s+", " ", `Legal Description`)]
dtAM[,LEGAL_DESCRIPTION := gsub("Plan:", "Plan", LEGAL_DESCRIPTION)]
dtAM[,LEGAL_DESCRIPTION := gsub("Block:", "Blk", LEGAL_DESCRIPTION)]
dtAM[,LEGAL_DESCRIPTION := gsub("Lot:", "Lot", LEGAL_DESCRIPTION)]
dtAM[,maxYear:=max(`Assessment Year`),by=`Account Number`]
dtAM <- dtAM[`Assessment Year`==maxYear] # lot size current==historical for ~all accounts. FORMALIZE THIS

dtPM <- merge(dtP, dtAM, by="LEGAL_DESCRIPTION",all.x=TRUE, all.y=FALSE)
dtPM <- dtPM[isRowHouse | isDetached]

# Now merge with census tracts to get shares of different projects by tract, elasticities, and mean price/rent
dfT <- st_read("~/DropboxExternal/dataRaw/lct_000b21a_e/lct_000b21a_e.shp")
dfT <- st_transform(dfT, crs = 4326)
dfT <- st_make_valid(dfT)

# use LATITUDE and LONGITUDE from permit data to get census tract for each permit by making a sf object and doing a spatial join
dfP <- st_as_sf(dtP[!is.na(LONGITUDE)], coords = c("LONGITUDE", "LATITUDE"), crs = 4326)
dfP <- st_make_valid(dfP)
dfP <- st_join(dfP, dfT, join = st_within)
dtP <- as.data.table(dfP)
# get fraction that is detached vs row house by tract
dtShare <- dtP[YEAR>2023,.(N=.N, NDetached=sum(isDetached), NRowHouse=sum(isRowHouse)), by=.(CTNAME)]

# same tract merge for the current-year assessor data
dfAM <- st_as_sf(dtAM[!is.na(Longitude)], coords = c("Longitude", "Latitude"), crs = 4326)
dfAM <- st_make_valid(dfAM)
dfAM <- st_join(dfAM, dfT, join = st_within)
dtAM <- as.data.table(dfAM)
dtAM[,assessed:=gsub("\\$","",`Assessed Value`)]
dtAM[,assessed:=as.numeric(gsub(",","",assessed))]

regHedonic <- feols(log(as.numeric(assessed)) ~ log(as.numeric(lot_size)) + log(`Total Gross Area`) | CTNAME, data=dtAM)
dtHedonic <- as.data.table(fixef(regHedonic), keep.rownames = TRUE)
names(dtHedonic) <- c("CTNAME","FE")

# elasticity by tract: cov(log(`Total Gross Area`), log(assessed)) / var(log(`Total Gross Area`))
dtAM[,lotNumeric := as.numeric(lot_size)]
dtAM <- dtAM[!is.na(lotNumeric) & !is.na(`Total Gross Area`) & !is.na(assessed) & lotNumeric>0 & `Total Gross Area`>0 & assessed>0]
dtCoef <- dtAM[,.(Elasticity=cov(log(`Total Gross Area`), log(assessed))/var(log(`Total Gross Area`))), by=.(CTNAME)]
setkey(dtCoef, "CTNAME")
setkey(dtHedonic, "CTNAME")
dtCoef <- merge(dtCoef, dtHedonic, by="CTNAME", all.x=TRUE, all.y=FALSE)
print(cor(dtCoef[,Elasticity], dtCoef[,FE], use="complete.obs"))
ggplot(dtCoef, aes(x=Elasticity, y=FE)) + geom_point() + geom_smooth(method="lm") + theme_bw() + labs(title="Elasticity vs Fixed Effect by Census Tract", x="Elasticity", y="Fixed Effect")
ggsave("text/edmontonElasticityFE.png")

setkey(dtShare, "CTNAME")
dtCoefShare <- dtCoef[dtShare, nomatch=0]
print(cor(dtCoefShare[,.(Elasticity, FE, NRowHouse/(NDetached+NRowHouse))], use="complete.obs"))
ggplot(dtCoefShare, aes(x=Elasticity, y=NRowHouse/(NDetached+NRowHouse))) + geom_point() + geom_smooth(method="lm") + theme_bw() + labs(title="Elasticity vs Share of Row Houses by Census Tract", x="Elasticity", y="Share of Row Houses")
ggsave("text/edmontonElasticityShare.png")
ggplot(dtCoefShare, aes(x=FE, y=NRowHouse/(NDetached+NRowHouse))) + geom_point() + geom_smooth(method="lm") + theme_bw() + labs(title="Fixed Effect vs Share of Row Houses by Census Tract", x="Fixed Effect", y="Share of Row Houses")
ggsave("text/edmontonFEShare.png")
q("no")

File: scripts/edmontonMatch.py
# edmontonMatch.py
# Match building permits to assessment
# Strategy: do 2026 and late 2025 permits so old address still valid maybe
# Tom Davidoff
# 07/16/26

#library(data.table)
#library(fixest)
#library(sf)

import polars as pl
import sys
import polars as pl

path = "/Users/tdavidof/DropboxExternal/dataRaw/edmonton/Property_Assessment_Data_(Historical)_20260611.csv"

dfA = (
    pl.scan_csv(path, infer_schema_length=0)
    .with_columns(
        (pl.col("Zoning") == "RS").any().over("Account Number").alias("maxRS")
    )
    .filter(pl.col("maxRS"))
)

result = dfA.collect()
print(result.head())
sys.exit()

# convert next 3 lines to polars
dfA = pl.read_csv("~/DropboxExternal/dataRaw/edmonton/Property_Assessment_Data_(Historical)_20260611.csv",infer_schema_length=0)
dfC = pl.read_csv("~/DropboxExternal/dataRaw/edmonton/Property_Information_(Current_Calendar_Year)_20260612.csv",infer_schema_length=0)
dfP = pl.read_csv("~/DropboxExternal/dataRaw/edmonton/edmontonBuildingPermits.csv",infer_schema_length=0)


# limit to current RS zones
dfA = dfA.with_columns((pl.col("Zoning")=="RS").any().over("Account Number").alias("maxRS"))
print(dfA.head())
# print a table of maxRS by Zoning
vs_table = dfA.pivot_count(on="maxRS", index="Zoning")
print(vs_table)
sys.exit()
dfdfd

#dtA[,table(maxRS==1,Zoning)]

# header rows
#==> /Users/tdavidof/DropboxExternal/dataRaw/edmonton/edmontonBuildingPermits.csv <==
#Row ID,PERMIT_DATE,PERMIT_NUMBER,REPORT_PERMIT_DATE,YEAR,MONTH_NUMBER,JOB_CATEGORY,JOB_DESCRIPTION,BUILDING_TYPE,WORK_TYPE,CONSTRUCTION_VALUE,FLOOR_AREA,UNITS_ADDED,ADDRESS,LEGAL_DESCRIPTION,ZONING,NEIGHBOURHOOD_NUMBER,NEIGHBOURHOOD,BIA,COUNT,LATITUDE,LONGITUDE,LOCATION,Geometry Point,Occupancy Date
#1-253374239,2019-07-04,,2019-07-04,2019,7,Commercial Final,To construct an LRT Station (Davies Station) on City Road Right-of-Way,Transportation Terminals (440),(01) New,20150000,39815,0,100 - DAVIES LRT STATION NW,"","",6170,DAVIES INDUSTRIAL WEST,"",1,,,"","",
#
#==> /Users/tdavidof/DropboxExternal/dataRaw/edmonton/Property_Assessment_Data_(Historical)_20260611.csv <==
#"Account Number","Assessment Year","Suite","House Number","Street Name","Legal Description","Latitude","Longitude","Point Location","Neighbourhood","Actual Year Built","Garage","Zoning","Lot Size","Assessed Value","Assessment Class 1","Assessment Class % 1","Assessment Class 2","Assessment Class % 2","Assessment Class 3","Assessment Class % 3"
#"1477850","2013",,"10821","150 STREET NW","Plan: 704KS  Block: 68  Lot: 5","53.5533","-113.57936","POINT (-113.57936 53.5533)","HIGH PARK","1955","Y","RF1","681.1240","$269,500.00","RESIDENTIAL","100",,,,
#
#==> /Users/tdavidof/DropboxExternal/dataRaw/edmonton/Property_Information_(Current_Calendar_Year)_20260612.csv <==
#"Account Number","Suite","House Number","Street Name","legal_description","zoning","lot_size","Total Gross Area","year_built","garage","Neighbourhood ID","Neighbourhood","Ward","Latitude","Longitude","Point Location"
#"11219634",,"303","29 STREET SW","Plan: 2322574  Block: 11  Lot: 42","RSF","335.364","183.4","2024","true","6669","ALCES","Sspomitapi","53.4300755348","-113.38586836927","POINT (-113.38586836927416 53.43007553480231)"

## first question: Suites in RS 2025+ completions
#print(head(dtC))
#dtCs <- dtC[zoning=="RS" & year_built>=2025 ]
#print(dtCs[,.(N=.N,hasSuite=sum(Suite!=""))])
# Answer: no suites, that's not how rowhouse for sale would be identified

# edmontonMatch_skeleton.R
# Structure for the three-file linkage. NOT runnable — stubs mark where
# real logic (canon(), knn match, demo chain, area-conservation) plugs in.
# Division of labour:
#   dtP (permits)      = CLASSIFIER  -> built form, units, suites
#   dtA (historical)   = PARENT      -> pre-redev lot_size (2023 vintage)
#   dtC (current year) = CHILD       -> new post-redev parcels (for assemblies)
#
# Column provenance (what each join produces):
#   builtForm, unitsAdded, suiteFlag   <- dtP        (never from roll)
#   parentLotSize, parentValue         <- dtA        via permit->parent join
#   n_parents, parent_area (assembly)  <- dtC<->dtA  child->parent join
#   builtForm MUST ride onto both the parent-lot number AND the assembly number

